File: drills/ts/solutions/nc105_word_search_ii__each_word.ts
export function findWords(board: string[][], words: string[]): string[] {
  if (board.length === 0 || board[0].length === 0) return [];

  // Word Search, once per word. Correct, and it redoes the search for every
  // shared prefix: a hundred words beginning "ab" each re-walk that "ab" from
  // every square. That repetition is exactly what the trie removes.
  const exists = (word: string): boolean => {
    if (word === "") return false;

    const walk = (r: number, c: number, at: number, used: Set<string>): boolean => {
      if (r < 0 || r >= board.length || c < 0 || c >= board[0].length) return false;
      if (used.has(`${r},${c}`) || board[r][c] !== word[at]) return false;
      if (at === word.length - 1) return true;
      const next = new Set(used);
      next.add(`${r},${c}`);
      return [[1, 0], [-1, 0], [0, 1], [0, -1]].some(([dr, dc]) =>
        walk(r + dr, c + dc, at + 1, next),
      );
    };

    for (let r = 0; r < board.length; r++) {
      for (let c = 0; c < board[0].length; c++) {
        if (walk(r, c, 0, new Set())) return true;
      }
    }
    return false;
  };

  return words.filter(exists);
}

File: drills/go/solutions/nc105_word_search_ii__each_word.go
package main

func findWords(board [][]byte, words []string) []string {
	// Run the single-word search once per word: correct, and simple, at
	// the cost of rewalking the board for every word.
	found := []string{}
	seen := map[string]bool{}
	for _, w := range words {
		if !seen[w] && exists(board, w) {
			seen[w] = true
			found = append(found, w)
		}
	}
	return found
}

func exists(board [][]byte, word string) bool {
	var search func(r, c, i int) bool
	search = func(r, c, i int) bool {
		if i == len(word) {
			return true
		}
		if r < 0 || r >= len(board) || c < 0 || c >= len(board[0]) || board[r][c] != word[i] {
			return false
		}
		saved := board[r][c]
		board[r][c] = '#'
		found := search(r+1, c, i+1) || search(r-1, c, i+1) || search(r, c+1, i+1) || search(r, c-1, i+1)
		board[r][c] = saved
		return found
	}
	for r := range board {
		for c := range board[r] {
			if search(r, c, 0) {
				return true
			}
		}
	}
	return false
}
